x-pack/legacy/plugins/infra/public/containers/logs/log_filter/index.ts

```diff
@@ -4,23 +4,114 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-import { useContext } from 'react';
+import { useReducer, useMemo } from 'react';
 import createContainer from 'constate';
-import { ReduxStateContext } from '../../../utils/redux_context';
-import { logFilterSelectors as logFilterReduxSelectors } from '../../../store/local/selectors';
+import { esKuery } from '../../../../../../../../src/plugins/data/public';
 
-export const useLogFilterState = () => {
-  const { local: state } = useContext(ReduxStateContext);
-  const filterQuery = logFilterReduxSelectors.selectLogFilterQueryAsJson(state);
-  return { filterQuery };
-};
+enum Action {
+  SetDraftFilterQuery,
+  ApplyFilterQuery,
+}
+
+interface SetDraftQueryAction {
+  type: Action.SetDraftFilterQuery;
+  payload: FilterQuery | null;
+}
+
+interface ApplyFilterQueryAction {
+  type: Action.ApplyFilterQuery;
+  payload: SerializedFilterQuery;
+}
+
+type ActionObj = SetDraftQueryAction | ApplyFilterQueryAction;
 
-export interface LogFilterStateParams {
-  filterQuery: string | null;
+export interface KueryFilterQuery {
+  kind: 'kuery';
+  expression: string;
 }
 
-export const logFilterInitialState = {
+export type FilterQuery = KueryFilterQuery;
+
+export interface SerializedFilterQuery {
+  query: FilterQuery;
+  serializedQuery: string;
+}
+
+interface LogFilterBaseStateParams {
+  filterQuery: SerializedFilterQuery | null;
+  filterQueryDraft: KueryFilterQuery | null;
+}
+
+export const logFilterInitialState: LogFilterBaseStateParams = {
   filterQuery: null,
+  filterQueryDraft: null,
+};
+
+interface LogFilterDerivatives {
+  isFilterQueryDraftValid: boolean;
+  filterQueryAsJson: SerializedFilterQuery['serializedQuery'] | null;
+  filterQuery: FilterQuery | null;
+}
+
+export type LogFilterStateParams = Omit<LogFilterBaseStateParams, 'filterQuery'> &
+  LogFilterDerivatives;
+
+export interface LogFilterCallbacks {
+  setLogFilterQueryDraft: (payload: FilterQuery) => void;
+  applyLogFilterQuery: (payload: SerializedFilterQuery) => void;
+}
+
+const getDerivatives = (state: LogFilterBaseStateParams) => ({
+  get isFilterQueryDraftValid() {
+    const { filterQueryDraft } = state;
+    if (filterQueryDraft && filterQueryDraft.kind === 'kuery') {
+      try {
+        esKuery.fromKueryExpression(filterQueryDraft.expression);
+      } catch (err) {
+        return false;
+      }
+    }
+
+    return true;
+  },
+  get filterQueryAsJson() {
+    const { filterQuery } = state;
+    return filterQuery ? filterQuery.serializedQuery : null;
+  },
+  get filterQuery() {
+    const { filterQuery } = state;
+    return filterQuery ? filterQuery.query : null;
+  },
+});
+
+export const useLogFilterState: () => [LogFilterStateParams, LogFilterCallbacks] = () => {
+  const [state, dispatch] = useReducer(logFilterStateReducer, logFilterInitialState);
```

Comment:
Given that the two pieces of the state are only read and written independently, how about we simplify this to two `useState`s and a few `useMemo`s? Something like

```typescript
const [logFilterQuery, setLogFilterQuery] = useState<FilterQuery | null>(null);
const [logFilterQueryDraft, setLogFilterQueryDraft] = useState<FilterQuery | null>(null);

const isFilterQueryDraftValid = useMemo(() => {
  if (logFilterQueryDraft && logFilterQueryDraft.kind === 'kuery') {
    try {
      esKuery.fromKueryExpression(logFilterQueryDraft.expression);
    } catch (err) {
      return false;
    }
  }
  return true;
}, [esKuery.fromKueryExpression, logFilterQueryDraft]);

const serializedLogFilterQuery = useMemo(() => {/*...*/}, [esKuery.fromKueryExpression, logFilterQuery]);
```

Reply:
`applyLogFilterQuery` does need to modify both properties at the same time, but I can work with this

Reply:
Why is `esKuery.fromKueryExpression` a dependency in these?